### PR TITLE
Enhance the main lexer benchmark.

### PR DESCRIPTION
**Pull request by chandlerc**
This generalizes the main lexer benchmark's source generation to be
a bit more comprehensive, specifically including whitespace and
comments. Without these, we're missing a key part of the lexer's
performance.

This also tidies up a bit of the code and adds a more specific
distribution of the different factors in lexing based on analysis of
LLVM's source code. The enhancements to the source statistics script
that helped collect the data here will be in a separate PR.

With this, the output on my AMD cloud instance is:
```
-------------------------------------------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations bytes_per_second tokens_per_second
-------------------------------------------------------------------------------------------------------------------------
BM_ValidKeywords                               2809011 ns      2808912 ns          243       212.616M/s         35.601M/s
BM_ValidIdentifiers<1, 64, false>             11461783 ns     11461652 ns           61       128.499M/s        8.72475M/s
BM_ValidIdentifiers<1, 1, true>                3397293 ns      3397240 ns          208       84.2158M/s        29.4357M/s
BM_ValidIdentifiers<3, 5, true>               14165143 ns     14164962 ns           51       40.3956M/s        7.05967M/s
BM_ValidIdentifiers<3, 16, true>              15283128 ns     15282583 ns           47       71.7623M/s         6.5434M/s
BM_ValidIdentifiers<12, 64, true>             17417323 ns     17417109 ns           41       219.007M/s        5.74148M/s
------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations bytes_per_second lines_per_second tokens_per_second
------------------------------------------------------------------------------------------------------------------------------------------
BM_RandomSource                                8132211 ns      8132227 ns           84       137.223M/s        3.9036M/s        12.2968M/s
BM_SpeedOfLightStrCpy                            29610 ns        29608 ns        24631       36.8065G/s       1072.17M/s        3.37745G/s
BM_SpeedOfLightDispatch<1>                     2144418 ns      2144421 ns          327       520.387M/s       14.8035M/s        46.6326M/s
BM_SpeedOfLightDispatch<2>                     1945954 ns      1945827 ns          351       573.498M/s       16.3144M/s         51.392M/s
BM_SpeedOfLightDispatch<4>                     2519565 ns      2519467 ns          292       442.923M/s       12.5999M/s        39.6909M/s
BM_SpeedOfLightDispatch<8>                     3011965 ns      3011968 ns          238       370.498M/s       10.5396M/s        33.2009M/s
BM_SpeedOfLightDispatch<16>                    4379575 ns      4379579 ns          160       254.803M/s       7.24841M/s        22.8332M/s
BM_SpeedOfLightDispatch<32>                    6678423 ns      6678353 ns          102       167.096M/s       4.75342M/s        14.9738M/s
BM_SpeedOfLightDispatch<MaxDispatchTargets>    9373075 ns      9372688 ns           75       119.062M/s       3.38697M/s        10.6693M/s
```

I've compared the profile of the `BM_RandomSource` benchmark with this
change and it largely corresponds to what I expect based on profiling
hand-crafted Carbon inputs. And as you can see, we're closing in on
lexing at least hitting the 10-million-lines-per-second mark. =]

### CODE REVIEW

File: toolchain/lex/tokenized_buffer_benchmark.cpp [hunk 1]
```diff
@@ -226,15 +226,66 @@ auto GetSymbolTokenTable() -> llvm::ArrayRef<TokenKind> {
...
+                  string_literal_percent) <= 100);
+
+    CARBON_CHECK(tokens_per_line <= NumTokens);
+    CARBON_CHECK(NumTokens % tokens_per_line == 0)
+        << "Tokens per line of " << tokens_per_line
+        << " does not divide the number of tokens " << NumTokens;
+
+    CARBON_CHECK(0 <= comment_line_percent && comment_line_percent <= 100);
+    CARBON_CHECK(0 <= blank_line_percent && blank_line_percent <= 100);
+    CARBON_CHECK((comment_line_percent + blank_line_percent) < 100);
```
Comment (zygoloid) [inline toolchain/lex/tokenized_buffer_benchmark.cpp]: I wonder if a helper function would help:
```suggestion
    auto is_percentage = [](int n) { return 0 <= n && n <= 100; };
    CARBON_CHECK(is_percentage(symbol_percent));
    CARBON_CHECK(is_percentage(keyword_percent));
    CARBON_CHECK(is_percentage(numeric_literal_percent));
    CARBON_CHECK(is_pecentage(string_literal_percent));
    CARBON_CHECK(is_percentage(symbol_percent + keyword_percent +
                               numeric_literal_percent + string_literal_percent));

    CARBON_CHECK(tokens_per_line <= NumTokens);
    CARBON_CHECK(NumTokens % tokens_per_line == 0)
        << "Tokens per line of " << tokens_per_line
        << " does not divide the number of tokens " << NumTokens;

    CARBON_CHECK(is_percentage(comment_line_percent));
    CARBON_CHECK(is_percentage(blank_line_percent));
    CARBON_CHECK(comment_line_percent + blank_line_percent < 100);
```
A comment on the last line explaining why the sum must be < 100% would have helped me here. (I see below that the reason is we need some lines to put the tokens on!)

**Comment by chandlerc** [comment]
(all dependencies merged)

File: toolchain/lex/tokenized_buffer_benchmark.cpp [hunk 2]
```diff
@@ -268,7 +319,48 @@ auto RandomMixedSeq(int symbol_percent, int keyword_percent) -> std::string {
...
+  // tokens per line rather than using some distribution with this as the median
+  // or mean.
+  llvm::SmallVector<std::string> lines;
+  // First place tokens onto each line.
+  for (auto i : llvm::seq(NumTokens / options.tokens_per_line)) {
+    lines.push_back("");
+    llvm::raw_string_ostream os(lines.back());
+    // Arbitrarily indent each line by two spaces.
+    os << "  ";
+    llvm::ListSeparator sep;
```
Comment (zygoloid) [inline toolchain/lex/tokenized_buffer_benchmark.cpp]: I assume we don't want commas between the tokens:
```suggestion
    llvm::ListSeparator sep(" ");
```

**Comment by chandlerc** [comment]
Thanks, all applied and updated as suggested, PTAL.

**Review by zygoloid — APPROVED** [review]

**Pull request merged**